### ISSUE TITLE
counsel-imenu-goto pass wrong argument to imenu

### ISSUE THREAD

**Issue by danyuhao** [redguardtoo/emacs.d#470]
I have modified the `counsel-imenu-goto`  function like this:
```
(defun counsel-imenu-goto ()
  "Go to buffer position"
  (interactive)
  (let ((imenu-auto-rescan t) items)
    (unless (featurep 'imenu)
      (require 'imenu nil t))
    (setq items (imenu--make-index-alist t))
    (setq items (delete (assoc "*Rescan*" items) items))
    (ivy-read "imenu items:"
              (ivy-imenu-get-candidates-from items)
              :action (lambda (k) (imenu k)))))

```
And I get the error debug info：
```
Debugger entered--Lisp error: (wrong-type-argument listp #<marker at 699245 in mmx_core.c>)
  imenu(("mmx_core_set_eqam_channel_annex" "mmx_core_set_eqam_channel_annex" . #<marker at 699245 in mmx_core.c>))
  (lambda (k) (imenu k))(("mmx_core_set_eqam_channel_annex" "mmx_core_set_eqam_channel_annex" . #<marker at 699245 in mmx_core.c>))
  ivy-call()
  ivy-read("imenu items:" (("mmx_core_init_by_phase" "mmx_core_init_by_phase" . #<marker at 12343 in mmx_core.c>) ("mmx_core_init" "mmx_core_init" . #<marker at 12917 in mmx_core.c>) ("mmx_core_eqam_data_init" "mmx_core_eqam_data_init" . #<marker at 35929 in mmx_core.c>) ("mmx_core_snr_data_init" "mmx_core_snr_data_init" . #<marker at 37582 in mmx_core.c>) ("mmx_core_snr_data_free" "mmx_core_snr_data_free" . #<marker at 37976 in mmx_core.c>) ("mmx_core_snr_data_copy" "mmx_core_snr_data_copy" . #<marker at 38316 in mmx_core.c>) ("mmx_core_udc_group_rule_data_init" "mmx_core_udc_group_rule_data_init" . #<marker at 38734 in mmx_core.c>) ("mmx_core_udc_group_rule_data_copy" "mmx_core_udc_group_rule_data_copy" . #<marker at 39373 in mmx_core.c>) ("mmx_core_udc_group_data_init" "mmx_core_udc_group_data_init" . #<marker at 40006 in mmx_core.c>) ("mmx_core_udc_group_data_free" "mmx_core_udc_group_data_free" . #<marker at 40764 in mmx_core.c>) ("mmx_core_udc_group_data_copy" "mmx_core_udc_group_data_copy" . #<marker at 41229 in mmx_core.c>) ("mmx_core_thread_init" "mmx_core_thread_init" . #<marker at 41716 in mmx_core.c>) ("mmx_core_deinit" "mmx_core_deinit" . #<marker at 42147 in mmx_core.c>) ("mmx_core_free_info" "mmx_core_free_info" . #<marker at 42512 in mmx_core.c>) ("mmx_core_run" "mmx_core_run" . #<marker at 52652 in mmx_core.c>) ("mac_timer_msg_proc" "mac_timer_msg_proc" . #<marker at 53962 in mmx_core.c>) ("mac_main_msg_proc" "mac_main_msg_proc" . #<marker at 59390 in mmx_core.c>) ("mac_fid_init_fun" "mac_fid_init_fun" . #<marker at 60071 in mmx_core.c>) ("mmx_core_set_rcc_template_channel" "mmx_core_set_rcc_template_channel" . #<marker at 60962 in mmx_core.c>) ("mmx_core_set_rcc_template_rc" "mmx_core_set_rcc_template_rc" . #<marker at 62349 in mmx_core.c>) ("mmx_core_set_rcc_template_rm" "mmx_core_set_rcc_template_rm" . #<marker at 63400 in mmx_core.c>) ("mmx_core_set_rcpctrlset" "mmx_core_set_rcpctrlset" . #<marker at 64456 in mmx_core.c>) ("mmx_core_set_cable_rf_attr" "mmx_core_set_cable_rf_attr" . #<marker at 65101 in mmx_core.c>) ("mmx_core_set_pre_eq_switch" "mmx_core_set_pre_eq_switch" . #<marker at 70296 in mmx_core.c>) ("mmx_core_get_pre_eq_enable" "mmx_core_get_pre_eq_enable" . #<marker at 70983 in mmx_core.c>) ("mmx_core_set_ip_mode" "mmx_core_set_ip_mode" . #<marker at 71314 in mmx_core.c>) ("mmx_core_set_cable_rf_switch" "mmx_core_set_cable_rf_switch" . #<marker at 71854 in mmx_core.c>) ("mmx_core_set_cmts_macaddr" "mmx_core_set_cmts_macaddr" . #<marker at 72777 in mmx_core.c>) ("mmx_core_get_rf_power_monitor" "mmx_core_get_rf_power_monitor" . #<marker at 73200 in mmx_core.c>) ("mmx_core_set_rf_adjust_freq_att" "mmx_core_set_rf_adjust_freq_att" . #<marker at 74958 in mmx_core.c>) ("mmx_core_set_rf_adjust_micro_adjust" "mmx_core_set_rf_adjust_micro_adjust" . #<marker at 75818 in mmx_core.c>) ("mmx_core_set_rf_adjust_temperature" "mmx_core_set_rf_adjust_temperature" . #<marker at 76710 in mmx_core.c>) ("mmx_core_report_rf_adjust_ds_chan_power" "mmx_core_report_rf_adjust_ds_chan_power" . #<marker at 77286 in mmx_core.c>) ("mmx_core_clear_ku115_xaui0_state" "mmx_core_clear_ku115_xaui0_state" . #<marker at 78969 in mmx_core.c>) ("mmx_core_set_load_balance_enable" "mmx_core_set_load_balance_enable" . #<marker at 79429 in mmx_core.c>) ("mmx_core_set_general_load_balance_enable" "mmx_core_set_general_load_balance_enable" . #<marker at 79958 in mmx_core.c>) ("mmx_core_set_restrict_load_balance_enable" "mmx_core_set_restrict_load_balance_enable" . #<marker at 80852 in mmx_core.c>) ("mmx_core_set_load_balance_init_tech" "mmx_core_set_load_balance_init_tech" . #<marker at 81396 in mmx_core.c>) ("mmx_core_add_or_detete_load_balance_group" "mmx_core_add_or_detete_load_balance_group" . #<marker at 82179 in mmx_core.c>) ("mmx_core_add_or_delete_chan_to_load_balance_group" "mmx_core_add_or_delete_chan_to_load_balance_group" . #<marker at 83535 in mmx_core.c>) ("mmx_core_add_or_detele_cm_to_load_balance_group" "mmx_core_add_or_detele_cm_to_load_balance_group" . #<marker at 84938 in mmx_core.c>) ("mmx_core_get_channel_cm_cnt" "mmx_core_get_channel_cm_cnt" . #<marker at 86294 in mmx_core.c>) ("mmx_core_get_cm_cnt_in_group" "mmx_core_get_cm_cnt_in_group" . #<marker at 87675 in mmx_core.c>) ("mmx_core_is_group_existed" "mmx_core_is_group_existed" . #<marker at 90361 in mmx_core.c>) ("mmx_core_get_all_cm_cnt_in_restrict_group" "mmx_core_get_all_cm_cnt_in_restrict_group" . #<marker at 90979 in mmx_core.c>) ("mmx_core_get_cm_cnt_in_genaral" "mmx_core_get_cm_cnt_in_genaral" . #<marker at 91855 in mmx_core.c>) ("mmx_core_get_cm_cnt_summary" "mmx_core_get_cm_cnt_summary" . #<marker at 94446 in mmx_core.c>) ("mmx_core_calc_cm_cnt_in_channel" "mmx_core_calc_cm_cnt_in_channel" . #<marker at 95998 in mmx_core.c>) ("mmx_core_is_load_group_member" "mmx_core_is_load_group_member" . #<marker at 97153 in mmx_core.c>) ("mmx_core_set_load_balance_param" "mmx_core_set_load_balance_param" . #<marker at 98294 in mmx_core.c>) ...) :action (lambda (k) (imenu k)))
  (let ((imenu-auto-rescan t) items) (if (featurep (quote imenu)) nil (require (quote imenu) nil t)) (setq items (imenu--make-index-alist t)) (setq items (delete (assoc "*Rescan*" items) items)) (ivy-read "imenu items:" (ivy-imenu-get-candidates-from items) :action (function (lambda (k) (imenu k)))))
  counsel-imenu-goto()
  call-interactively(counsel-imenu-goto nil nil)
  command-execute(counsel-imenu-goto)
```
It seems that the `imenu` get the wrong list argument `("mmx_core_set_eqam_channel_annex" "mmx_core_set_eqam_channel_annex" . #<marker at 699245 in mmx_core.c>) `???

And when I just type` M-x imenu RET mmx_core_set_eqam_channel_annex`,   it can jump to the right place

**Comment by redguardtoo** [comment]
what's your major-mode?

**Comment by danyuhao** [comment]
I worked in c mode, C-h m got the info
```
C/lhw mode defined in `cc-mode.el':
Major mode for editing K&R and ANSI C code.
```
The problem is resolved after  I  copied the `counsel-imenu` funtion to replace `counsel-imenu-goto` from  [abo's config](https://github.com/abo-abo/swiper/blob/master/counsel.el). And I notice that it use `(imenu (cdr candidate))` instead of (imenu k)`. Maybe it is the point??

**Comment by redguardtoo** [comment]
That's already fixed in e21c89d error handling in imenu based on ivy (Chen Bin).

BTW, abo's code was actually written by me.

**Comment by danyuhao** [comment]
Great job!
